File: src/main/java/ltdjms/discord/aiagent/services/ToolExecutionInterceptor.java
package ltdjms.discord.aiagent.services;

import java.lang.reflect.Array;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.Instant;
import java.util.Collection;
import java.util.HexFormat;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import javax.inject.Inject;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.fasterxml.jackson.databind.ObjectMapper;

import ltdjms.discord.aiagent.domain.ToolExecutionLog;
import ltdjms.discord.aiagent.persistence.ToolExecutionLogRepository;
import ltdjms.discord.shared.events.DomainEventPublisher;
import ltdjms.discord.shared.events.LangChain4jToolExecutedEvent;
import ltdjms.discord.shared.events.LangChain4jToolExecutionStartedEvent;

/**
 * 工具執行審計攔截器。
 *
 * <p>記錄所有 LangChain4J 工具調用到 tool_execution_log 表，並發布事件通知。 審計紀錄只保存去敏摘要，不保存原始參數、回傳內容或錯誤內容。
 *
 * <p>使用 ThreadLocal 存儲當前執行上下文，當工具執行完成時記錄審計日誌。
 */
public final class ToolExecutionInterceptor {

  private static final Logger LOGGER = LoggerFactory.getLogger(ToolExecutionInterceptor.class);

  private final ToolExecutionLogRepository logRepository;
  private final ObjectMapper objectMapper;
  private final DomainEventPublisher eventPublisher;
  private final ThreadLocal<ExecutionContext> context = new ThreadLocal<>();

  /**
   * 建立工具執行審計攔截器。
   *
   * @param logRepository 工具執行日誌 Repository
   * @param objectMapper JSON 序列化器
   * @param eventPublisher 領域事件發布器
   */
  @Inject
  public ToolExecutionInterceptor(
      ToolExecutionLogRepository logRepository,
      ObjectMapper objectMapper,
      DomainEventPublisher eventPublisher) {
    this.logRepository = logRepository;
    this.objectMapper = objectMapper;
    this.eventPublisher = eventPublisher;
  }

  /**
   * 記錄工具執行開始。
   *
   * @param toolName 工具名稱
   * @param parameters 工具參數
   */
  public void onToolExecutionStarted(String toolName, Map<String, Object> parameters) {
    try {
      ToolExecutionContext.Context ctx = ToolExecutionContext.getContext();
      context.set(
          new ExecutionContext(
              ctx.guildId(),
              ctx.channelId(),
              ctx.userId(),
              toolName,
              parameters,
              System.currentTimeMillis()));

      LangChain4jToolExecutionStartedEvent event =
          new LangChain4jToolExecutionStartedEvent(
              ctx.guildId(), ctx.channelId(), ctx.userId(), toolName, Instant.now());
      eventPublisher.publish(event);
    } catch (IllegalStateException e) {
      LOGGER.warn("無法獲取工具執行上下文，跳過審計記錄");
    }
  }

  /**
   * 記錄工具執行成功並發布事件。
   *
   * @param result 執行結果
   * @return 工具執行通知訊息
   */
  public String onToolExecutionCompleted(String result) {
    ExecutionContext ctx = context.get();
    if (ctx == null) {
      LOGGER.debug("無工具執行上下文，跳過成功記錄");
      return result;
    }

    try {
      String parametersJson = summarizeParameters(ctx.parameters());
      String resultSummary = summarizeTextPayload(result);
      ToolExecutionLog log =
          ToolExecutionLog.success(
              ctx.guildId(),
              ctx.channelId(),
              ctx.userId(),
              ctx.toolName(),
              parametersJson,
              resultSummary);
      logRepository.save(log);

      // 發布工具執行事件
      LangChain4jToolExecutedEvent event =
          new LangChain4jToolExecutedEvent(
              ctx.guildId(),
              ctx.channelId(),
              ctx.userId(),
              ctx.toolName(),
              result,
              true,
              Instant.now());
      eventPublisher.publish(event);

      LOGGER.debug("工具執行審計已記錄: tool={}, success", ctx.toolName());

      // 返回工具執行通知訊息
      return "✅ 工具「" + getToolDisplayName(ctx.toolName()) + "」執行成功";

    } catch (Exception e) {
      LOGGER.error("記錄工具執行成功日誌失敗: tool={}", ctx.toolName(), e);
      return result;
    } finally {
      context.remove();
    }
  }

  /**
   * 記錄工具執行失敗並發布事件。
   *
   * @param error 錯誤訊息
   * @return 工具執行通知訊息
   */
  public String onToolExecutionFailed(String error) {
    ExecutionContext ctx = context.get();
    if (ctx == null) {
      LOGGER.debug("無工具執行上下文，跳過失敗記錄");
      return "❌ 工具執行失敗：" + error;
    }

    try {
      String parametersJson = summarizeParameters(ctx.parameters());
      String errorSummary = summarizeTextPayload(error);
      ToolExecutionLog log =
          ToolExecutionLog.failure(
              ctx.guildId(),
              ctx.channelId(),
              ctx.userId(),
              ctx.toolName(),
              parametersJson,
              errorSummary);
      logRepository.save(log);

      // 發布工具執行失敗事件
      LangChain4jToolExecutedEvent event =
          new LangChain4jToolExecutedEvent(
              ctx.guildId(),
              ctx.channelId(),
              ctx.userId(),
              ctx.toolName(),
              error,
              false,
              Instant.now());
      eventPublisher.publish(event);

      LOGGER.debug("工具執行審計已記錄: tool={}, failed", ctx.toolName());

      // 返回工具執行失敗通知訊息
      return "❌ 工具「" + getToolDisplayName(ctx.toolName()) + "」執行失敗：" + error;

    } catch (Exception e) {
      LOGGER.error("記錄工具執行失敗日誌失敗: tool={}", ctx.toolName(), e);
      return "❌ 工具執行失敗：" + error;
    } finally {
      context.remove();
    }
  }

  /**
   * 獲取工具的顯示名稱。
   *
   * @param toolName 工具名稱
   * @return 顯示名稱
   */
  private String getToolDisplayName(String toolName) {
    return switch (toolName) {
      case "createChannel" -> "創建頻道";
      case "createCategory" -> "創建類別";
      case "createRole" -> "創建角色";
      case "listChannels" -> "列出頻道";
      case "listCategories" -> "列出類別";
      case "listRoles" -> "列出角色";
      case "getChannelPermissions" -> "獲取頻道權限";
      case "getRolePermissions" -> "獲取角色權限";
      case "modifyChannelSettings", "modifyChannelPermissions" -> "修改頻道設定";
      case "modifyCategorySettings", "modifyCategoryPermissions" -> "修改類別設定";
      case "modifyRoleSettings", "modifyRolePermissions" -> "修改角色設定";
      default -> toolName;
    };
  }

  private String summarizeParameters(Map<String, Object> parameters) {
    if (parameters == null || parameters.isEmpty()) {
      return "{\"redacted\":true,\"entryCount\":0,\"keys\":[]}";
    }

    Map<String, Object> summary = new LinkedHashMap<>();
    Map<String, Object> valueSummaries = new LinkedHashMap<>();
    List<String> keys = parameters.keySet().stream().sorted().toList();

    summary.put("redacted", true);
    summary.put("entryCount", parameters.size());
    summary.put("keys", keys);

    for (String key : keys) {
      valueSummaries.put(key, summarizeValue(parameters.get(key)));
    }
    summary.put("values", valueSummaries);

    return toJson(summary, "{\"redacted\":true,\"entryCount\":0,\"keys\":[]}");
  }

  private Map<String, Object> summarizeValue(Object value) {
    Map<String, Object> summary = new LinkedHashMap<>();
    summary.put("redacted", true);

    if (value == null) {
      summary.put("type", "null");
      return summary;
    }

    summary.put("type", value.getClass().getSimpleName());

    if (value instanceof CharSequence text) {
      summary.put("length", text.length());
      summary.put("blank", text.toString().isBlank());
    } else if (value instanceof Collection<?> collection) {
      summary.put("size", collection.size());
    } else if (value instanceof Map<?, ?> map) {
      summary.put("size", map.size());
    } else if (value.getClass().isArray()) {
      summary.put("size", Array.getLength(value));
    }

    summary.put("sha256", sha256Hex(safeValueFingerprint(value)));
    return summary;
  }

  private String summarizeTextPayload(String payload) {
    Map<String, Object> summary = new LinkedHashMap<>();
    summary.put("redacted", true);
    summary.put("type", "text");
    summary.put("length", payload == null ? 0 : payload.length());
    summary.put("blank", payload == null || payload.isBlank());
    summary.put("sha256", sha256Hex(payload == null ? "" : payload));
    return toJson(summary, "{\"redacted\":true,\"type\":\"text\"}");
  }

  private String safeValueFingerprint(Object value) {
    try {
      return objectMapper.writeValueAsString(value);
    } catch (Exception e) {
      LOGGER.warn("無法序列化工具參數摘要來源，改用字串雜湊", e);
      return String.valueOf(value);
    }
  }

  private String toJson(Object value, String fallback) {
    try {
      return objectMapper.writeValueAsString(value);
    } catch (Exception e) {
      LOGGER.warn("轉換工具審計摘要為 JSON 失敗", e);
      return fallback;
    }
  }

  private String sha256Hex(String value) {
    try {
      MessageDigest digest = MessageDigest.getInstance("SHA-256");
      byte[] hash = digest.digest(value.getBytes(StandardCharsets.UTF_8));
      return HexFormat.of().formatHex(hash);
    } catch (NoSuchAlgorithmException e) {
      throw new IllegalStateException("SHA-256 not available", e);
    }
  }

  /**
   * 工具執行上下文數據。
   *
   * @param guildId 伺服器 ID
   * @param channelId 頻道 ID
   * @param userId 用戶 ID
   * @param toolName 工具名稱
   * @param parameters 工具參數
   * @param startTime 開始時間
   */
  private record ExecutionContext(
      long guildId,
      long channelId,
      long userId,
      String toolName,
      Map<String, Object> parameters,
      long startTime) {}
}

File: packages/ai/src/services/ToolExecutionInterceptor.ts
import { AsyncLocalStorage } from 'node:async_hooks';
import { createHash } from 'node:crypto';
import pino from 'pino';
import type {
  DomainEventPublisher,
  LangChain4jToolExecutedEvent,
  LangChain4jToolExecutionStartedEvent,
} from '@ltdjms/shared';
import { ToolExecutionContext } from '../tools/ToolExecutionContext.js';
import {
  createFailureToolExecutionLog,
  createSuccessToolExecutionLog,
} from '../domain/tool-execution-log.js';
import type { ToolExecutionLogRepository } from '../persistence/drizzle-tool-execution-log-repository.js';

interface ExecutionContext {
  guildId: string;
  channelId: string;
  userId: string;
  toolName: string;
  parameters: Record<string, unknown> | null;
  startTime: number;
  managedByRun?: boolean;
}

const EMPTY_PARAMETERS_SUMMARY = '{"redacted":true,"entryCount":0,"keys":[]}';
const TEXT_FALLBACK_SUMMARY = '{"redacted":true,"type":"text"}';

/**
 * Intercepts tool execution for audit logging and domain events.
 * Matches Java ToolExecutionInterceptor (redacted params/results, DB + events).
 */
export class ToolExecutionInterceptor {
  private static readonly executionStorage = new AsyncLocalStorage<ExecutionContext>();

  private readonly logger: pino.Logger;
  private readonly forceJsonFailure: boolean;

  constructor(
    private readonly logRepository?: ToolExecutionLogRepository,
    private readonly eventPublisher?: DomainEventPublisher,
    logger?: pino.Logger,
    options?: { forceJsonFailure?: boolean },
  ) {
    this.logger = logger ?? pino({ name: 'tool-execution-interceptor' });
    this.forceJsonFailure = options?.forceJsonFailure ?? false;
  }

  /**
   * Runs a tool execution fn with isolated interceptor context (concurrency-safe).
   */
  async runTracked<T>(
    toolName: string,
    parameters: Record<string, unknown>,
    fn: () => Promise<T>,
  ): Promise<T> {
    const ctx = ToolExecutionContext.getContext();
    if (!ctx) {
      return fn();
    }

    const executionCtx: ExecutionContext = {
      guildId: ctx.guildId,
      channelId: ctx.channelId,
      userId: ctx.userId,
      toolName,
      parameters: parameters ?? null,
      startTime: Date.now(),
      managedByRun: true,
    };

    return ToolExecutionInterceptor.executionStorage.run(executionCtx, async () => {
      this.onToolExecutionStarted(toolName, parameters);
      try {
        const result = await fn();
        if (typeof result === 'string') {
          this.onToolExecutionCompleted(result);
        }
        return result;
      } catch (error) {
        this.onToolExecutionFailed(error);
        throw error;
      }
    });
  }

  onToolExecutionStarted(toolName: string, _parameters: Record<string, unknown>): void {
    try {
      const ctx = ToolExecutionContext.getContext();
      if (!ctx) {
        this.logger.warn('無法獲取工具執行上下文，跳過審計記錄');
        return;
      }

      if (!ToolExecutionInterceptor.executionStorage.getStore()) {
        this.logger.debug('無 runTracked 上下文，跳過工具開始審計');
        return;
      }

      if (this.eventPublisher) {
        const event: LangChain4jToolExecutionStartedEvent = {
          eventType: 'langchain4j_tool_execution_started',
          guildId: ctx.guildId,
          channelId: ctx.channelId,
          userId: ctx.userId,
          toolName,
          timestamp: new Date(),
        };
        this.eventPublisher.publish(event);
      }
    } catch (error) {
      this.logger.warn({ err: error }, 'Failed to record tool execution start');
    }
  }

  onToolExecutionCompleted(result: string): string {
    const ctx = ToolExecutionInterceptor.executionStorage.getStore();
    if (!ctx) {
      this.logger.debug('無工具執行上下文，跳過成功記錄');
      return result;
    }

    try {
      const parametersJson = this.summarizeParameters(ctx.parameters);
      const resultSummary = this.summarizeTextPayload(result);
      const log = createSuccessToolExecutionLog(
        ctx.guildId,
        ctx.channelId,
        ctx.userId,
        ctx.toolName,
        parametersJson,
        resultSummary,
      );

      void this.persistLog(log);

      if (this.eventPublisher) {
        const event: LangChain4jToolExecutedEvent = {
          eventType: 'langchain4j_tool_executed',
          guildId: ctx.guildId,
          channelId: ctx.channelId,
          userId: ctx.userId,
          toolName: ctx.toolName,
          result,
          success: true,
          timestamp: new Date(),
        };
        this.eventPublisher.publish(event);
      }

      return `✅ 工具「${this.getToolDisplayName(ctx.toolName)}」執行成功`;
    } catch (error) {
      this.logger.error({ err: error, toolName: ctx.toolName }, '記錄工具執行成功日誌失敗');
      return result;
    } finally {
      this.clearExecutionContext();
    }
  }

  onToolExecutionFailed(error: unknown): string {
    const ctx = ToolExecutionInterceptor.executionStorage.getStore();
    const message = error instanceof Error ? error.message : String(error);

    if (!ctx) {
      this.logger.debug('無工具執行上下文，跳過失敗記錄');
      return `❌ 工具執行失敗：${message}`;
    }

    try {
      const parametersJson = this.summarizeParameters(ctx.parameters);
      const errorSummary = this.summarizeTextPayload(message);
      const log = createFailureToolExecutionLog(
        ctx.guildId,
        ctx.channelId,
        ctx.userId,
        ctx.toolName,
        parametersJson,
        errorSummary,
      );

      void this.persistLog(log);

      if (this.eventPublisher) {
        const event: LangChain4jToolExecutedEvent = {
          eventType: 'langchain4j_tool_executed',
          guildId: ctx.guildId,
          channelId: ctx.channelId,
          userId: ctx.userId,
          toolName: ctx.toolName,
          result: message,
          success: false,
          timestamp: new Date(),
        };
        this.eventPublisher.publish(event);
      }

      return `❌ 工具「${this.getToolDisplayName(ctx.toolName)}」執行失敗：${message}`;
    } catch (persistError) {
      this.logger.error({ err: persistError, toolName: ctx.toolName }, '記錄工具執行失敗日誌失敗');
      return `❌ 工具執行失敗：${message}`;
    } finally {
      this.clearExecutionContext();
    }
  }

  private clearExecutionContext(): void {
    // Context lifetime is scoped to runTracked(); no AsyncLocalStorage mutation here.
  }

  private async persistLog(log: ReturnType<typeof createSuccessToolExecutionLog>): Promise<void> {
    if (!this.logRepository) {
      return;
    }
    const result = await this.logRepository.save(log);
    if (result.isErr()) {
      this.logger.error({ err: result.getError() }, 'Failed to persist tool execution log');
    }
  }

  private getToolDisplayName(toolName: string): string {
    const displayNames: Record<string, string> = {
      create_channel: '創建頻道',
      createChannel: '創建頻道',
      create_category: '創建類別',
      createCategory: '創建類別',
      create_role: '創建角色',
      createRole: '創建角色',
      list_channels: '列出頻道',
      listChannels: '列出頻道',
      list_categories: '列出類別',
      listCategories: '列出類別',
      list_roles: '列出角色',
      listRoles: '列出角色',
      get_channel_permissions: '獲取頻道權限',
      getChannelPermissions: '獲取頻道權限',
      get_role_permissions: '獲取角色權限',
      getRolePermissions: '獲取角色權限',
      modify_channel_permissions: '修改頻道設定',
      modifyChannelPermissions: '修改頻道設定',
      modify_category_permissions: '修改類別設定',
      modifyCategoryPermissions: '修改類別設定',
      modify_role_permissions: '修改角色設定',
      modifyRolePermissions: '修改角色設定',
    };
    return displayNames[toolName] ?? toolName;
  }

  summarizeParameters(parameters: Record<string, unknown> | null): string {
    if (!parameters || Object.keys(parameters).length === 0) {
      return EMPTY_PARAMETERS_SUMMARY;
    }

    const keys = Object.keys(parameters).sort();
    const valueSummaries: Record<string, unknown> = {};
    for (const key of keys) {
      valueSummaries[key] = this.summarizeValue(parameters[key]);
    }

    return this.toJson(
      {
        redacted: true,
        entryCount: keys.length,
        keys,
        values: valueSummaries,
      },
      EMPTY_PARAMETERS_SUMMARY,
    );
  }

  summarizeTextPayload(payload: string | null | undefined): string {
    const text = payload ?? '';
    return this.toJson(
      {
        redacted: true,
        type: 'text',
        length: text.length,
        blank: text.trim().length === 0,
        sha256: this.sha256Hex(text),
      },
      TEXT_FALLBACK_SUMMARY,
    );
  }

  private summarizeValue(value: unknown): Record<string, unknown> {
    const summary: Record<string, unknown> = { redacted: true };

    if (value === null || value === undefined) {
      summary.type = 'null';
      return summary;
    }

    if (Array.isArray(value)) {
      summary.type = 'Array';
      summary.size = value.length;
    } else if (value instanceof Map) {
      summary.type = 'Map';
      summary.size = value.size;
    } else if (typeof value === 'object') {
      summary.type = 'Object';
      summary.size = Object.keys(value as object).length;
    } else {
      summary.type = typeof value;
      if (typeof value === 'string') {
        summary.length = value.length;
        summary.blank = value.trim().length === 0;
      }
    }

    summary.sha256 = this.sha256Hex(this.safeValueFingerprint(value));
    return summary;
  }

  private safeValueFingerprint(value: unknown): string {
    try {
      return JSON.stringify(value);
    } catch {
      return String(value);
    }
  }

  private toJson(value: unknown, fallback: string): string {
    if (this.forceJsonFailure) {
      return fallback;
    }
    try {
      return JSON.stringify(value);
    } catch {
      return fallback;
    }
  }

  private sha256Hex(value: string): string {
    return createHash('sha256').update(value, 'utf8').digest('hex');
  }
}
